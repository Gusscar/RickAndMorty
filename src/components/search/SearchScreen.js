import React, { useMemo } from "react";
import { useSelector } from "react-redux";
import { useNavigate, useLocation } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import queryString from "query-string";
import { CharactersCard } from "../characters/CharactersCard";

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);

  const { list } = useSelector((state) => state.characters);

  const [formValues, handleInputChange] = useForm({
    searchText: q,
  });

  const { searchText } = formValues;

  const getCharactersByName = (name = "") => {
    if (name === "") {
      return [];
    }
    name = name.toLowerCase();

    return list?.results?.filter(
      (item) =>
        item.name.toLowerCase().includes(name) ||
        item.species.toLowerCase().includes(name) ||
        item.gender.toLowerCase().includes(name) ||
        item.created.toLowerCase().includes(name)
    );
  };

  const characterFileted = useMemo(() => getCharactersByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Searchs</h1>
      <hr />

      <div
        className="row"
        style={{ display: "flex", justifyContent: "center" }}
      >
        <div className="col-4">
          <h4>Search Rick And Morty</h4>
          <hr />

          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Buscar un Heroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={handleInputChange}
            />

            <button className="btn btn-primary mt-3 btn-block" type="submit">
              Search...
            </button>
          </form>
        </div>

        <div className="col-4">
          <h4>Results</h4>
          <hr />
          {q === "" ? (
            <div className="alert alert-info">Search a Character</div>
          ) : (
            characterFileted?.length === 0 && (
              <div className="alert alert-danger">
                There are no Results: {q}
              </div>
            )
          )}

          {characterFileted?.map((item) => (
            <CharactersCard key={item.id} {...item} />
          ))}
        </div>
      </div>
    </>
  );
};
