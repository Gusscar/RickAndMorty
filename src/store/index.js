import { configureStore } from "@reduxjs/toolkit";

//reducer
import characters from "./slices/characters";
import locations from "./slices/location";

export default configureStore({
  reducer: {
    characters,
    locations,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
});
