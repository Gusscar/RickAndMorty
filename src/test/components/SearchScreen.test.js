import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../components/search/SearchScreen";
import store from "../../store";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("test in <SearchScreen/>", () => {
  test("should show correctly with default values", () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/search"]}>
          <SearchScreen />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toBe(
      "Search a Character"
    );
  });

  test("should show Rick and Input with the value QueryString", () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/search?q=Rick"]}>
          <SearchScreen />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("input").prop("value")).toBe("Rick");
  });

  test("should call the navigate new screen", () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/search"]}>
          <SearchScreen />
        </MemoryRouter>
      </Provider>
    );

    expect(
      wrapper.find("input").simulate("change", {
        target: {
          name: "searchText",
          value: "rick",
        },
      })
    );

    wrapper.find("form").prop("onSubmit")({
      preventDefault() {},
    });

    expect(mockNavigate).toHaveBeenCalledWith("?q=rick");
  });
});
