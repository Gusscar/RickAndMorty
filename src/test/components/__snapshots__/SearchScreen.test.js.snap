// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <SearchScreen/> should show correctly with default values 1`] = `
Array [
  <h1>
    Searchs
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Rick And Morty
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un Heroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-3 btn-block"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Character
      </div>
    </div>
  </div>,
]
`;
